crates/sui-json-rpc/src/metrics.rs

```diff
@@ -0,0 +1,161 @@
+use jsonrpsee::core::__reexports::serde_json;
+use prometheus::{
+    register_histogram_vec_with_registry, register_int_counter_vec_with_registry, HistogramVec,
+    IntCounterVec,
+};
+use serde::Deserialize;
+use tokio::time::Instant;
+use tower::Layer;
+
+#[derive(Debug, Clone)]
+pub struct MetricsLayer {
+    metrics: Arc<Metrics>,
+}
+impl MetricsLayer {
+    pub fn new(registry: &prometheus::Registry) -> Self {
+        let metrics = Arc::new(Metrics {
+            requests_by_route: register_int_counter_vec_with_registry!(
+                "rpc_requests_by_route",
+                "Number of requests by route",
+                &["route"],
+                registry,
+            )
+            .unwrap(),
+            req_latency_by_route: register_histogram_vec_with_registry!(
+                "req_latency_by_route",
+                "Latency of a request by route",
+                &["route"],
+                LATENCY_SEC_BUCKETS.to_vec(),
+                registry,
+            )
+            .unwrap(),
+            errors_by_route: register_int_counter_vec_with_registry!(
+                "errors_by_route",
+                "Number of errors by route",
+                &["route"],
+                registry,
+            )
+            .unwrap(),
+        });
+
+        Self { metrics }
+    }
+}
+
+impl<S> Layer<S> for MetricsLayer {
+    type Service = JsonRpcMetricService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        JsonRpcMetricService::new(inner, self.metrics.clone())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct JsonRpcMetricService<S> {
+    inner: S,
+    metrics: Arc<Metrics>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Metrics {
+    /// Counter of requests, route is a label (ie separate timeseries per route)
+    requests_by_route: IntCounterVec,
+    /// Request latency, route is a label
+    req_latency_by_route: HistogramVec,
+    /// Failed requests by route
+    errors_by_route: IntCounterVec,
+}
+
+const LATENCY_SEC_BUCKETS: &[f64] = &[
+    0.001, 0.005, 0.01, 0.05, 0.1, 0.25, 0.5, 1., 2.5, 5., 10., 20., 30., 60., 90.,
+];
+
+impl<S> JsonRpcMetricService<S> {
+    pub fn new(inner: S, metrics: Arc<Metrics>) -> Self {
+        Self { inner, metrics }
+    }
+}
+
+impl<S> Service<Request<Body>> for JsonRpcMetricService<S>
+where
+    S: Service<Request<Body>, Response = Response<Body>> + Clone + Send + 'static,
+    S::Response: 'static,
+    S::Error: Into<Box<dyn Error + Send + Sync>> + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = Box<dyn Error + Send + Sync + 'static>;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
```

Comment:
is this layer called before or after request deserialization? If it's before, it'd be helpful to record deser errors separately